iroh-rpc/src/stream.rs

```diff
@@ -1,96 +1,190 @@
 use std::collections::HashMap;
 use std::io::BufRead;
 use std::pin::Pin;
+use std::sync::{Arc, Mutex};
 
 use bytecheck::CheckBytes;
 use futures::channel::{mpsc, oneshot};
 use futures::prelude::*;
+use futures::select;
 use futures::stream::StreamExt;
-use futures::task::{Context, Poll};
+use futures::task::{AtomicWaker, Context, Poll};
 use libp2p::PeerId;
 use rkyv;
 use rkyv::{Archive, Deserialize, Serialize};
+use tokio::task::JoinHandle;
 use tracing::{debug, error};
 
 use crate::commands::Command;
 use crate::error::RpcError;
 
-/// InStream coordinates a stream of packet data, allowing the user
 /// to receive the chunks of bytes in the correct order.
 // TODO: better name, add timeout, early cancel, and handle errors
-pub struct InStream {
-    packet_receiver: StreamReceiver,
+pub struct OrderedStream {
     out_sender: mpsc::Sender<Command>,
-    chunks: HashMap<u64, Vec<u8>>,
+    error_receiver: mpsc::Receiver<RpcError>,
+    chunks: Arc<Mutex<HashMap<u64, Vec<u8>>>>,
     id: u64,
     next_chunk: u64,
     num_chunks: u64,
+    // Handle to the stream's task
+    task: Arc<Mutex<AtomicWaker>>,
+    // Handle to the poll receiver task
+    handle: JoinHandle<()>,
+    // shutdown receiver loop
+    early_shutdown: Option<oneshot::Sender<()>>,
 }
 
-impl InStream {
+impl OrderedStream {
     pub fn new(
         header: Header,
         packet_receiver: StreamReceiver,
         out_sender: mpsc::Sender<Command>,
     ) -> Self {
-        InStream {
-            packet_receiver,
+        let (error_sender, error_receiver) = mpsc::channel(8);
+
+        let (shutdown_sender, shutdown_receiver) = oneshot::channel();
+        let chunks = Arc::new(Mutex::new(HashMap::new()));
+        let task = Arc::new(Mutex::new(AtomicWaker::new()));
+        OrderedStream {
             out_sender,
+            error_receiver,
             next_chunk: 0,
             id: header.id,
             num_chunks: header.num_chunks,
-            chunks: HashMap::new(),
+            chunks: Arc::clone(&chunks),
+            task: Arc::clone(&task),
+            handle: {
+                tokio::spawn(async move {
+                    OrderedStream::gather_packets(
+                        packet_receiver,
+                        chunks,
+                        task,
+                        error_sender,
+                        shutdown_receiver,
+                    )
+                    .await
+                })
+            },
+            early_shutdown: Some(shutdown_sender),
         }
     }
+
+    // we have loop that is taking in packets and ordering them. If there is an
+    // error, send the error to the main stream
+    async fn gather_packets(
+        mut packet_receiver: StreamReceiver,
+        chunks: Arc<Mutex<HashMap<u64, Vec<u8>>>>,
```

Comment:
These `Mutex`s must be async, as you are calling them in async functions, use `tokio::sync::Mutex`. Currently all `.lock()` calls will block the eventloop

Reply:
!!!!!!

Reply:
thank you

Reply:
What about what it's saying here:
https://docs.rs/tokio/latest/tokio/sync/struct.Mutex.html#which-kind-of-mutex-should-you-use

> The feature that the async mutex offers over the blocking mutex is the ability to keep it locked across an .await point ... The primary use case for the async mutex is to provide shared mutable access to IO resources such as a database connection. If the value behind the mutex is just data, it’s usually appropriate to use a blocking mutex such as the one in the standard library or [parking_lot](https://docs.rs/parking_lot).

Since in all cases that we are using a `lock` we use it and then immediately drop it (and not holding it "across an `.await` point), it seems like we can use `std::sync::mutex`

Basically, I'm hoping this is the case, since we can't `await` in `poll_next`

I've tested it out using a 4MB file and it seems to alternate between the `gather_packets` loop and then `poll_next`.

Reply:
interesting, I guess it is okay to use like this here